### PR TITLE
[api] Relocate `intf` modules according to dependency-order.

### PR DESCRIPTION
**Pull request by ejgallego**
In a component-based source code organization of Coq `intf` doesn't
fit very well, as it sits in bit of "limbo" between different
components, and indeed, encourages developers to place types in
sometimes random places wrt the hierarchy. For example, lower parts of
the system reference `Vernacexpr`, which morally lives in a pretty
higher part of the system.

We move all the files in `intf` to the lowest place their dependencies
can accommodate:

- `Misctypes`: is used by Declaremod, thus it has to go in `library` or below. Ideally, this file would disappear.
- `Evar_kinds`: it is used by files in `engine`, so that seems its proper placement.
- `Decl_kinds`: used in `library`, seems like the right place. [could also be merged]
- `Glob_term`: belongs to pretyping, where it is placed.
- `Locus`: ditto.
- `Pattern`: ditto.
- `Genredexpr`: depended by a few modules in `pretyping`, seems like the righ place.
- `Constrexpr`: used in `pretyping`, the use is a bit unfortunate and could be fixed, as this module should be declared in `interp` which is the one eliminating it.
- `Vernacexpr`: same problem than `Constrexpr`; this one can be fixed as it contains stuff it shouldn't. The right place should be `parsing`.
- `Extend`: Is placed in `pretyping` due to being used by `Vernacexpr`.
- `Notation_term`: First place used is `interp`, seems like the right place.

Additionally, for some files it could be worth to merge files of the form `Foo` with `Foo_ops` in the medium term, as to create proper ADT modules as done in the kernel with `Name`, etc...

Depends on #6859

### CODE REVIEW

File: META.coq [hunk 1]
```diff
@@ -9,26 +9,26 @@
 #  linkopts(byte)   = "-dllpath ${coq_install_path}/kernel/byterun/ -dllib -lcoqrun"
 
 description = "The Coq Proof Assistant Plugin API"
-version = "8.7"
+version = "8.8"
```
Comment (SkySkimmer) [inline META.coq]: This version number change seems unrelated.

Comment (ejgallego) [reply META.coq]: I thought it was a good moment to bump as `META` indeed has non-backwards-compatible changes in this PR.

**Comment by herbelin** [comment]
A few years ago, I was told that `intf` was good, now I read that it encourages bad practices. Nothing wrong in itself as it may be a standard thesis, antithesis, synthesis process, but just wanting to be sure that we were remaining consistent with respect to the global knowledge and motivations of developers.

> developers like to reference high-level components such as Vernacexpr from lower parts of the system.

As far as I'm concerned, there are no deep reasons for that. If the first time a type (e.g. `Vernacexpr.typeclass_constraint` or `Vernacexpr.locality_flag`) is needed is in `Vernacexpr`, it shall be defined in `Vernacexpr`. If the first time it occurs, it is in an other file, it shall be defined in this other file and `Vernacexpr` shall then refer to it. This being said, as far as I'm concerned, I have no particular opinion on where `Vernacexpr` should be. I can adapt to the cost of a move if this move is necessary.

**Comment by ejgallego** [comment]
Indeed @herbelin I consider this PRs more of a start of a discussion [with a proof of concept to check where this concrete path leads] than a strong defense of this model. It would be good if someone could remind the positive sides of the current `mli`-only setup.

Most of the changes stem from my branch using `jbuilder`, tool which enforces strict [maybe _too_ strict] modularity policies. And indeed, `mli`-only files seem to have some particular limitations specially in the context of modern "module aliased" OCaml.

Maybe a rule of thumb is "do we want the datatype definitions and the datatype operations to live separate from each other?" However, I do think that `intf` is a bit of a strange animal today as it packs together very unrelated files.

**Comment by herbelin** [comment]
> It would be good if someone could remind the positive sides of the current mli-only setup.

@letouzey could probably tell more about it, but the gain was at some time in reducing (a lot) the graph of dependencies, and in particular that the cma for printers or parsers had to draw (a lot of) files with them only because the files they were depending on for the types had also functions which were drawing other files, needed for the code but unneeded from the strict point of view of dependency in types.

As far as I understand, your current changes do not increase the complexity of the dependency graph. So, with regards to this, it respects the original objective of `intf`. However, the principle "the datatype definitions and the datatype operations (have) to live (next to) each other" may have the drawback mentioned above.

I agree however that `intf` is a bit of a strange animal. So, maybe a good compromise would be to keep such files collecting types (so that depending on these types do not draw uselessly a chain of dependencies which is needed only for the body of functions over these types) but that this type-only files are put in directories which are conceptually relevant for them.

**Comment by ejgallego** [comment]
@herbelin I am not sure I understand what you mean by "chain of dependencies" here, why was this a problem for printer and parsers?

[A different question is why printers and parsers cannot be defined in the corresponding module, but that seems more delicate indeed]

**Comment by herbelin** [comment]
> @herbelin I am not sure I understand what you mean by "chain of dependencies" here, why was this a problem for printer and parsers?

At some time (did it change with recent ocaml?), if a file `A.ml` was needing a type in `B.ml`/`B.mli`, I think to remember that the `B.cmo` was needed at linking time to build a `cma` for `A`. So, if the `B.ml` depended on a `C.ml`, even if `C.ml` was irrelevant for `B.mli`, the `C.cmo` had to be added as a dependency too. I'm not very competent on this but that is my memories.

Then, at some time, we had `printers.cma` (for debugging), and `grammar.cma` (for parsing), and they were drawing much more dependencies than really needed, and it was a pain to find a correct order, sometimes needing hacks. The `intf` drastically reduces the size of dependencies for these `cma`, iirc.

> [A different question is why printers and parsers cannot be defined in the corresponding module, but that seems more delicate indeed]

You mean the printer for term in `constr.ml` etc.? One answer is that printing a term requires some technology (notations, implicit arguments, scopes, ...). But maybe you meant something else.

**Comment by maximedenes** [comment]
WG discussion: backtracking on previous design, but seems ok since 1- permits better modular view 2- the main types and operations the user is expected to work on were already not there. @ppedrot says @letouzey is also not against such a change. Is that right, @letouzey?

**Comment by letouzey** [comment]
@maximedenes : for the record, I **do** object to this PR and the destruction of mli-only interfaces.
I will elaborate more on that and the mli-only approach later, but for now the main point is the following:

This `intf`directory regroups datatypes forming various IR for Coq. In IR, the I stands for Intermediate, right ? So no @ejgallego , your PR definitively does not put these files at the "proper place". For instance, It's just as wrong to put `Glob_term` in `pretyping` than in `interp`, since this IR is precisely between these components. Conversely, several components are conversions between two IRs.
The `intf/*.mli` approach is hence a way to highlight these IRs, and make them live *above* the components. And I still see that as clean and neat.

**Comment by maximedenes** [comment]
@letouzey Any chance you can attend the WG this afternoon so that we can discuss it?

**Comment by ejgallego** [comment]
> This `intf` directory regroups datatypes forming various IR for Coq. In IR, the I stands for Intermediate, right ? So no @ejgallego , your PR definitively does not put these files at the "proper place". For instance, It's just as wrong to put Glob_term in pretyping than in interp, since this IR is precisely between these components.

I think the last sentence is incorrect. `pretying` can work stand-alone regardless of `interp`, so if you see them as separate components `glob_term` does indeed belong in `pretyping`. It is not "in between". In fact, I wonder what does "in between" mean. I cannot assign a concrete meaning to it.

> Conversely, several components are conversions between two IRs.

And? The same applies to `Constr` then, why is it not in `intf`?

> The intf/*.mli approach is hence a way to highlight these IRs, and make them live above the components. And I still see that as clean and neat.

I guess it depends if you want to see Coq as a modular system or not. If you take the modular view, `intf` is nothing but a spaguetti. What is `Decl_kinds` doing next to the AST?

`intf` has also a very dangerous effect wrt code organization, as it fully breaks modularity. This way, no so long ago in the codebase we had components from `library` using objects from the Ast. This is clearly wrong if you want to follow a modular view.

As for type-only `.ml` files, my experience is that they are very confusing for plugin developers. Here is a large AST, but hey, look, no operations!

I still fail to see what is the point of "highlighting" an AST, when you have no idea how to use them, what is the component that defines/operates over it, etc...

**Comment by mattam82** [comment]
I kind of understand both points of view, one that intf summarizes the ILs in a clean, mli-only way, and the fact that indeed it also becomes a little bit of pot-pourri at the same time (missing EConstr and Constr) as well. I agree with @letouzey that they are interfaces between components (think compiler phases), and deciding to put the language def in the phase that produces it or consumes it is a bit arbitrary. Maybe there is a third way, putting this intf files in separate directories, separate from the phases that consume/produce them, but with the functions that manipulate them (views, constructors, substitutions etc), would be a good way to solve this?

**Comment by mattam82** [comment]
Also @maximedenes put forward the idea at the WG that the intf directory is only a small step towards having a global view of the types that make the different phases, and a document (latex or otherwise)  compiling the corresponding files would be just as useful (and could include things like Constr.t as well easily).

**Comment by ejgallego** [comment]
> the functions that manipulate them (views, constructors, substitutions etc), would be a good way to solve this

That would be good, however the interesting manipulation functions do indeed tend to depend on the component that uses them. For example, for `glob_term` you would end up with basically with `pretyping`. I don't see a case where `glob_term` is useful alone without the pretyper.

But indeed if a more refined "interface" organization is needed you can use module aliases as other projects do to present a coherent `glob_term` interface, however note that all the cases I know such interfaces include both types and functions. So we could indeed have:

```ocaml
module Intf : sig
  include Kernel_intf
  include Pretyping_intf
  include Interp_intf
```
etc...

**Comment by ejgallego** [comment]
About `mli` only files, there is this detail that makes life "interesting". Let's take Coq 8.6, which used `mli` file for `intf`. Now let's take all the `.cma` produced in a Coq build. In which `cma` is the module `Constrexpr`, for example?

**Comment by maximedenes** [comment]
So @mattam82 what should I do with this PR?

**Comment by ejgallego** [comment]
Note that if we want more discussion this PR can be perfectly postponed to 8.9. All-in-all, the choices here are about tradeoffs, so now that other parts taking advantage of the a more modular code structured a starting to be ready maybe we can value the gains/losses better in a next WG with concrete use case examples.

**Comment by ejgallego** [comment]
#6869 highlights a problem of `mli`-only modules: they cannot be packed, thus the interfaces land in the global namespace which a problem [for example your code is now not compatible with packing strategies based on aliases]

**Comment by maximedenes** [comment]
Ok, so I personally support the removal of the `intf` folder, as it is more confusing than anything else. In particular, not having there the main type on which we recommend plugin writers to work is a dealbreaker IMHO.

But as the release manager, I feel that the discussion hasn't converged yet, so I'll label this again as needing discussion.

**Comment by ejgallego** [comment]
So I'd like to take this PR to solve all the issues we are having in the short term with `intf` and `mli`-only files. Right now, the status of the tree is a mess in that sense, with some files floating around in different form and lots of "I like `mli` files" but little concern about the problems they do bring.

As you all remember, the can of worms was opened by #652 which had to remove the `.mli` files as to be able to alias interfaces. This is problem **1**, `mli` files cannot be aliases and so far nobody has proposed a solution other than using a `.ml` file.

The second problem #652 didn't fully solved but was completed later by #808 and #905 was indeed to eliminate cyclic dependencies from the Coq codebase. This is problem **2** and directly related to this PR, as indeed `API.mli` didn't follow the `intf` structure but a dependency based one.

Even if we didn't keep the resulting `.mli` file, the API effort was extremely positive and as of today we are close to a fully modular codebase barring the existance of `intf`, which encourages routine layering violations that should become clear once the merge queue allows us to reach the commits.

All in all, we need to complete the work started by API and provide a fully modular, "packable" codebase so we can follow modern recommended practice in OCaml building, which among other things say:

- module packs are built with aliases [thus every module needs an implementation]
- module interfaces and implementation should be considered a single unit [thus barring circular dependencies].
- libraries are packaged by functionality.

Of course I'd be happy to adapt any alternative solution that we may consider. Hey, we could even keep this directory, but of what use would be for the plugin builder to import `coq.intf` What gain I do have having in scope all kinds of unrelated scope if I only want to read and process `.vo` files.

**Comment by ejgallego** [comment]
> I'm lost in the details of this discussion...

Indeed, sorry @herbelin, the whole endeavor since API was started has been kinda exploratory so it is obviously very difficult to track. I'll try summarize:

1. For code organization reasons, we'd like to be able to pack and alias OCaml modules. Currently, this cannot be done unless the module has an implementation.
2. Similarly, for code organization reasons we'd like to make every Coq directory an ocamlfind unit, so we have `coq.kernel`, `coq.engine` etc...

So **2** is almost there, however **1** interacts subtly with it. The moment things in *1* get an implementation dependencies don't respect the directory boundaries. Example, `Vernacexpr` depends on things in library, so `coq.intf` has to depend on `coq.library`, however, `coq.library` also depends on other things on `coq.intf`. So at the _package_ level, the dependency we have is cyclical.

Why it works today? Only because our module system sees the global picture and computes a correct ordering regardless of the "directory" boundaries. You can see this by looking at the order make does process files. So this remains a barrier to truly modularize Coq or building it with `dune` which is composable [the build will be global] but requires modularity [as to able to split components off transparently]

> Are the approaches "automatic generation of the ml from the mli", and "symbolic links" or are there others?

IMHO "automatic generation of the ml from the mli" is a bit of a hack an prone to errors as now `.mli` files must be explicitly declared [doing so by default is dangerous as imagine the case you rename a file, etc...] and it adds cumbersome rules. "symbolic links" would be great, but they don't work on windows.

Regarding the API documentation, I think `odoc` et al shouldn't really care about that. Not that we are using the documentation so much but I hope `jbuilder build @doc` convinces people to improve it a bit.

**Comment by herbelin** [comment]
@ejgallego: Well, I think I understand what you want to do and this seems legitimate. But are there no alternative which would preserve the other expectations? After all, we already generate `.ml` files in binary format from `.ml4` files. We could maybe generate `.ml` files in binary format from `.mli` files (or is it a stupid idea)? About the dependencies issues with `intf`, would splitting `intf` into two parts an alternative solution (or, again, is it a stupid idea)?

**Comment by ejgallego** [comment]
> We could maybe generate .ml files in binary format from .mli files (or is it a stupid idea)?

I am wary of this as you have no mechanism to detect whether you are in an error for example due to renaming. By the way, no mechanism for generating `.ml` files from `.mli` works in all cases [yeah, contrary to belief you cannot copy and `.mli` and it will be a valid one, but I think that Coq won't fall into that corner case, see some discussion at https://github.com/ocaml/dune/issues/9)

A possible transition strategy would indeed be "move to .ml files", and resurrect if we adopt Dune so every interested person can add a manual declaration. But yeah, this is a bit of a maintenance burden but at least is kinda supported [even if it suffers from some pitfalls as it is really a hack]

What Théo has proposed [mli files really do make no sense, as types should appear together with the operations for documentation purposes] is another sensible alternative, but longer term.

> About the dependencies issues with intf, would splitting intf into two parts an alternative solution (or, again, is it a stupid idea)?

If you look at the dependency chain it is more fine grained than that [API showed it right](https://github.com/coq/coq/blob/f6857ce53ecf64b0086854495b4a8451f476d5b4/API/API.mli)
look at where the files do occur in that large .mli file.

So in fact, I think that the files in `intf` are morally spread in the following manner:
- `Extend` + `Constrexpr` + `Vernacexpr` do belong to `parsing`, as it is the main producer. Would we split the parser they would be also split. #6515 does indeed implement this.
- `Decl_kinds` belongs to `library`
- `Evar_kinds` is undoubtedly part of `engine`
- `Locus` `Pattern` `Genredexpr` and `Glob_term` are core datatypes of `pretyper`.
- `Notation_term` is indeed owned by `interp`
- `Misctypes` should IMHO go away, as its name indicates, it is a _melange_ of different types, and most of them would be better off somewhere.

The above placement has been validated by a strictly modular build tool.

**Comment by ejgallego** [comment]
All in all, it boils down to whether we consider that the ideas in 929d25a05585dd702739b6979e3822bfa6cdbadb and its parents have stood the test of time.

IMHO, that kind of splitting was not very positive as the previous combined modules seemed easier to understand and _document_. Now, it is hard to keep the interface files in sync with the corresponding `_ops` ones, and I'd dare to say that this had led to quite a few redundancy in plugins and the codebase, as `_ops` are not easy to locate. I'm all for undoing the `_ops` split, and I think some other people think likewise.

I could well missing somethings as I still fail to see the motivation/advantages of the split, I tried to deduce it from the logs but so far without success.

**Comment by ejgallego** [comment]
Hi all, so I am going to propose this revised PR for merge. see the updated description, but all that the PR does is to move files to the component that they do belong, if we see the source as collection of components.

This PR is essential in getting Coq to build with a modular tool such as Dune, and IMHO it also solves an interesting problem in the new merge model: declaring a maintainer for `intf` is an oxymoron as `intf` is really a collection of bits from several components [for example changes to `glob_constr` should be approved by the pretyper person].

Let me know if you have any further question.

**Comment by ejgallego** [comment]
Also, I think that discussion about `mli` vs `ml` files should continue in #6664 ; I think that choice is fairly orthogonal to this PR as the `mli/ml` problem appears in a few components in addition to `intf`.

File: META.coq [hunk 2]
```diff
@@ -20,7 +20,7 @@ package "clib" (
   version     = "8.8"
 
   directory   = "clib"
-  requires    = "str, unix, threads"
+  requires    = "num, str, unix, threads"
```
Comment (SkySkimmer) [inline META.coq]: How's this change related to this PR?

Comment (ejgallego) [reply META.coq]: Likely was needed for proper testing; let's split to its corresponding OR.

**Review by Zimmi48 — APPROVED** [review]

**Comment by herbelin** [comment]
Quick comments:
- About the question of why you are proposing to have `vernacexpr`and `constrexpr` with their producer (in `parsing`), and `glob_term` with its consumer (in `pretyping`), rather than the converse choices, and rather than in specific stand-alone directories, I'm still not convinced of the superiority of a solution over the others. In particular, even if I would not fight against an arbitrarily-looking choice to be made, I still don't see why using Dune would be easier with one solution than with the other or why reviewing would be clearer with one solution than with the other.
- For the purpose of grepping and clear separation between code and interface, it is ok to me if there are `ml` files with a canonical name of the form `foo_intf.ml`.
- No particular opinion on the `_ops`, status quo or merge are both ok (see a similar question at item 3 [here](https://github.com/coq/coq/pull/7109#issue-178267055)).

**Comment by ejgallego** [comment]
> For the purpose of grepping and clear separation between code and interface, it is ok to me if there are ml files with a canonical name of the form foo_intf.ml.

I like the naming, however I am not sure about what is the exact meaning we'd like to "code" and "interface", as I wrote in #6664 sometimes the line is blurry [also in coq with `_ops.ml` modules] I suggest we continue the discussión at #6664.

> About the question of why you are proposing to have vernacexprand constrexpr with their producer (in parsing), and glob_term with its consumer (in pretyping), rather than the converse choices, and rather than in specific stand-alone directories, I'm still not convinced of the superiority of a solution over the others. In particular, even if I would not fight against an arbitrarily-looking choice to be made, I still don't see why using Dune would be easier with one solution than with the other or why reviewing would be clearer with one solution than with the other.

Sorry if the explanation is not very good; in the modular view, each directory is "packed" into a top-level module and forms a component. Thus, mixing files that belong to different modules in the same directory is going to be a problem, specially if as today would imply cyclic dependencies. The best way to understand it is in fact try to use Dune for example and move a few files around; you quickly see how the dependency graph is subtle. With regards to the consumer/producer distinction, even if not enforced here, datatypes should always belong to the one processing them. In this sense, you don't have a lot of choice; `parsing` cannot be placed below `interp` so `constrexpr` belongs to `interp`, unless you place parsing below. For `vernac`, it is placed above `parsing` thus you have to put the vernacular AST in the parser, or move the parser above `vernac` which is a bit complicated [you can also make it a free-floating component, but not without significant refactoring I think.

**Comment by ejgallego** [comment]
Note that this is not the only option, other "relocations" are possible. For example, we could keep most of the `intf` directory if we place it _after_ `library`, and do a bit of surgery on `Misctypes` to move `module_kind`. Also, in most solutions, I'll be hard not to place extend with `cLexer`.

I am fine with any solution that removes cyclic dependencies when modules are considered as a single compilation unit.

**Review by gares — APPROVED** [review]

**Comment by gares** [comment]
I think this PR is a clear step forward.

**Review by maximedenes — COMMENTED** [review]
The changes in `.github/CODEOWNERS` are obviously ok.

**Review by maximedenes — APPROVED** [review]

**Review by herbelin — COMMENTED** [review]
Summary of previous comments: There are different solutions to go in the direction of a component-based organization of the code. Among these solutions, the solution proposed in this PR is ok to me. If other solutions have to be considered to get a general agreement, these other solutions, as soon as they preserve the different objectives we have in mind, shall be ok to me.

In any case, let's not delay too much the discussion, if ever more discussion is needed.

**Review by ppedrot — APPROVED** [review]
I don't have a strong opinion on this PR, but I don't want delay merging it, so this is a green light from me.

**Comment by gares** [comment]
I'm the secondary maintainer for `pretyping/` IIRC, so if @mattam82 is busy, here my thoughts.

IMO the changes to pretyping are OK, and I totally agree with @ejgallego that `vernacexpr` should be in parsing, but moving it there requires some more work that should not interfere with this PR.

**Comment by ejgallego** [comment]
Yeah I think we can finally move on on this.

**Comment by Zimmi48** [comment]
@ppedrot I think you finally got all the green lights you were requesting, didn't you?

**Comment by ejgallego** [comment]
Thanks! A few rebases will be needed [sorry for that]

**Pull request merged**